core/node/state_keeper/src/io/mempool.rs

```diff
@@ -286,6 +288,45 @@ impl StateKeeperIO for MempoolIO {
         Ok(())
     }
 
+    async fn rollback_l2_block(&mut self, txs: Vec<Transaction>) -> anyhow::Result<()> {
+        let mut to_add = Vec::with_capacity(txs.len());
+        for tx in txs
+            .into_iter()
+            .filter(|tx| tx.tx_format() != TransactionType::ProtocolUpgradeTransaction)
+            .rev()
+        {
+            let constraint = self.mempool.rollback(&tx);
+            to_add.push((tx, constraint));
+        }
+
+        to_add.reverse();
+        self.mempool.insert(to_add, HashMap::new());
+
+        Ok(())
+    }
+
+    async fn advance_nonces(&mut self, txs: Box<&mut (dyn Iterator<Item = &Transaction> + Send)>) {
```

Comment:
is it clear that `advance_nonces` also advances the `serial_id`? Maybe it is - you know better - but if not consider commenting/name change